File: custom_components/glt_flow_card/project_diff.py
"""Policy-driven semantic project diff and dependency closure."""
from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any, Mapping

from .project_contract import digest_canonical_json, evaluate_project_contract

_ROOT = Path(__file__).resolve().parents[2]
DIFF_POLICY = json.loads((_ROOT / "schemas" / "diff-policy.json").read_text(encoding="utf-8"))
_IDENTITY_COLLECTIONS = set(DIFF_POLICY["order"]["identity_keyed_collections"])
_MOVE_FIELDS = set(DIFF_POLICY["category_rules"]["move_fields"])
_MISSING = object()


def _utf16_sort_key(value: str) -> bytes:
    return value.encode("utf-16-be", errors="surrogatepass")


def _clone(value: Any) -> Any:
    return json.loads(digest_canonical_json(value)["canonical"])


def _pointer_part(value: Any) -> str:
    return str(value).replace("~", "~0").replace("/", "~1")


def _valid_document(raw_input: Any, label: str) -> tuple[dict[str, Any], dict[str, Any]]:
    evidence = evaluate_project_contract(raw_input)
    if not evidence["valid"]:
        details = ", ".join(f'{error["code"]}@{error["path"]}' for error in evidence["errors"])
        raise ValueError(f"{label} project contract is invalid: {details}")
    return json.loads(evidence["canonical"]), evidence


def _value_hash(value: Any, present: bool) -> str | None:
    return digest_canonical_json(value)["digest"] if present else None


def _category_for(parts: list[str]) -> str:
    last = parts[-1] if parts else ""
    if any(part in _MOVE_FIELDS for part in parts):
        return "move"
    if (
        last in {"entity", "entity_id", "state_entity"}
        or (parts and parts[0] == "bindings")
        or (len(parts) >= 3 and parts[0] == "fields" and last == "entity")
        or (len(parts) >= 3 and parts[0] == "slots" and last == "entity_id")
    ):
        return "binding"
    return DIFF_POLICY["category_rules"]["fallback_category"]


def _impact_for(category: str, path: str) -> dict[str, Any]:
    if path == "/security" or path.startswith("/security/") or path == "/permissions" or path.startswith(
        "/permissions/"
    ) or path == "/plugins" or path.startswith("/plugins/"):
        return {"severity": "critical", "areas": ["security"]}
    areas = {
        "add": ["none"],
        "remove": ["operational", "referential"],
        "move": ["visual"],
        "binding": ["binding", "operational"],
        "config": ["operational"],
    }[category]
    return {
        "severity": DIFF_POLICY["impact"]["default_by_category"][category],
        "areas": areas,
    }


def _operation(
    category: str,
    path: str,
    collection: str | None,
    object_id: str | None,
    relative_parts: list[str],
    before: Any,
    before_present: bool,
    after: Any,
    after_present: bool,
) -> dict[str, Any]:
    return {
        "id": f"{category}:{path}",
        "category": category,
        "path": path,
        "collection": collection,
        "object_id": object_id,
        "field": f'/{"/".join(_pointer_part(part) for part in relative_parts)}' if relative_parts else "",
        "before_hash": _value_hash(before, before_present),
        "after_hash": _value_hash(after, after_present),
        "before": _clone(before) if before_present else None,
        "after": _clone(after) if after_present else None,
        "impact": _impact_for(category, path),
        "requires": [],
    }


def _same_primitive(before: Any, after: Any) -> bool:
    if isinstance(before, bool) or isinstance(after, bool):
        return type(before) is type(after) and before == after
    if isinstance(before, (int, float)) and isinstance(after, (int, float)):
        return before == after
    return type(before) is type(after) and before == after


def _compare_value(
    operations: list[dict[str, Any]],
    before: Any,
    after: Any,
    *,
    path: str,
    collection: str | None,
    object_id: str | None,
    relative_parts: list[str],
) -> None:
    before_present = before is not _MISSING
    after_present = after is not _MISSING
    if before_present and after_present and not isinstance(before, (dict, list)) and not isinstance(after, (dict, list)):
        if _same_primitive(before, after):
            return
    if not before_present or not after_present:
        category = ("remove" if before_present else "add") if not relative_parts else _category_for(relative_parts)
        operations.append(
            _operation(category, path, collection, object_id, relative_parts, before, before_present, after, after_present)
        )
        return
    before_container = isinstance(before, (dict, list))
    after_container = isinstance(after, (dict, list))
    if type(before) is not type(after) or not before_container:
        operations.append(_operation(
            _category_for(relative_parts), path, collection, object_id, relative_parts, before, True, after, True
        ))
        return
    if isinstance(before, list):
        for index in range(max(len(before), len(after))):
            _compare_value(
                operations,
                before[index] if index < len(before) else _MISSING,
                after[index] if index < len(after) else _MISSING,
                path=f"{path}/{index}",
                collection=collection,
                object_id=object_id,
                relative_parts=[*relative_parts, str(index)],
            )
        return
    keys = sorted(set(before) | set(after), key=_utf16_sort_key)
    for key in keys:
        _compare_value(
            operations,
            before.get(key, _MISSING),
            after.get(key, _MISSING),
            path=f"{path}/{_pointer_part(key)}",
            collection=collection,
            object_id=object_id,
            relative_parts=[*relative_parts, key],
        )


def _identity_map(document: Mapping[str, Any], collection: str) -> dict[str, dict[str, Any]]:
    identity_field = DIFF_POLICY["identity_fields"][collection]
    return {entry[identity_field]: entry for entry in document.get(collection, [])}


def _add_dependencies(
    operations: list[dict[str, Any]], before: Mapping[str, Any], after: Mapping[str, Any]
) -> None:
    operation_ids = {operation["id"] for operation in operations}
    before_maps = {collection: _identity_map(before, collection) for collection in DIFF_POLICY["identity_fields"]}
    after_maps = {collection: _identity_map(after, collection) for collection in DIFF_POLICY["identity_fields"]}
    for operation in operations:
        collection = operation["collection"]
        object_id = operation["object_id"]
        if not collection or not object_id:
            continue
        references = [
            reference for reference in DIFF_POLICY["dependencies"]["references"] if reference["from"] == collection
        ]
        source_maps = before_maps if operation["category"] == "remove" else after_maps
        source = source_maps[collection].get(object_id)
        if source is None:
            continue
        requirements: dict[str, dict[str, str]] = {}
        for reference in references:
            for field in reference["fields"]:
                target_id = source.get(field)
                if not isinstance(target_id, str):
                    continue
                target_path = f'/{reference["to"]}/{_pointer_part(target_id)}'
                prefix = "remove" if operation["category"] == "remove" else "add"
                target_operation = f"{prefix}:{target_path}"
                if target_operation in operation_ids:
                    requirements[target_operation] = {
                        "operation_id": target_operation,
                        "reason": f'reference:{reference["from"]}.{field}->{reference["to"]}',
                    }
        operation["requires"] = sorted(requirements.values(), key=lambda value: _utf16_sort_key(value["operation_id"]))


def compute_project_diff(before_input: Any, after_input: Any) -> dict[str, Any]:
    """Return deterministic semantic operations for two validated projects."""

    before, before_evidence = _valid_document(before_input, "source")
    after, after_evidence = _valid_document(after_input, "candidate")
    operations: list[dict[str, Any]] = []
    ordering_noise: list[str] = []
    for key in sorted(set(before) | set(after), key=_utf16_sort_key):
        if key in _IDENTITY_COLLECTIONS and isinstance(before.get(key), list) and isinstance(after.get(key), list):
            before_map = _identity_map(before, key)
            after_map = _identity_map(after, key)
            before_ids = list(before_map)
            after_ids = list(after_map)
            if len(before_ids) == len(after_ids) and set(before_ids) == set(after_ids) and before_ids != after_ids:
                ordering_noise.append(f"/{_pointer_part(key)}")
            for object_id in sorted(set(before_ids) | set(after_ids), key=_utf16_sort_key):
                _compare_value(
                    operations,
                    before_map.get(object_id, _MISSING),
                    after_map.get(object_id, _MISSING),
                    path=f"/{_pointer_part(key)}/{_pointer_part(object_id)}",
                    collection=key,
                    object_id=object_id,
                    relative_parts=[],
                )
            continue
        _compare_value(
            operations,
            before.get(key, _MISSING),
            after.get(key, _MISSING),
            path=f"/{_pointer_part(key)}",
            collection=None,
            object_id=None,
            relative_parts=[key],
        )
    operations.sort(key=lambda operation: _utf16_sort_key(operation["id"]))
    _add_dependencies(operations, before, after)
    return {
        "policy_version": DIFF_POLICY["policy_version"],
        "source_digest": before_evidence["digest"],
        "candidate_digest": after_evidence["digest"],
        "operations": operations,
        "ordering_noise": sorted(ordering_noise, key=_utf16_sort_key),
    }


def expand_diff_selection(diff_result: Mapping[str, Any], selected_operation_ids: list[str]) -> dict[str, Any]:
    """Expand selected operations to deterministic transitive policy dependencies."""

    operations = {operation["id"]: operation for operation in diff_result.get("operations", [])}
    requested = sorted(set(selected_operation_ids or []), key=_utf16_sort_key)
    for operation_id in requested:
        if operation_id not in operations:
            raise ValueError(f"unknown selected operation {operation_id}")
    state: dict[str, str] = {}
    included: set[str] = set()
    added: dict[str, dict[str, str]] = {}

    def visit(operation_id: str, required_by: str | None = None, reason: str | None = None, chain=None) -> None:
        active_chain = list(chain or [])
        if operation_id not in operations:
            raise ValueError(f"missing dependency operation {operation_id}")
        if state.get(operation_id) == "visiting":
            raise ValueError(f'cyclic diff dependency: {" -> ".join([*active_chain, operation_id])}')
        if state.get(operation_id) == "done":
            return
        state[operation_id] = "visiting"
        if required_by is not None and operation_id not in requested and operation_id not in added:
            added[operation_id] = {
                "operation_id": operation_id,
                "required_by": required_by,
                "reason": reason or "",
            }
        requirements = sorted(
            operations[operation_id].get("requires", []), key=lambda value: _utf16_sort_key(value["operation_id"])
        )
        for requirement in requirements:
            visit(requirement["operation_id"], operation_id, requirement["reason"], [*active_chain, operation_id])
        state[operation_id] = "done"
        included.add(operation_id)

    for operation_id in requested:
        visit(operation_id)
    return {
        "selected": sorted(included, key=_utf16_sort_key),
        "requested": requested,
        "added": sorted(added.values(), key=lambda value: _utf16_sort_key(value["operation_id"])),
    }


def _json_lines() -> None:
    sys.stdin.reconfigure(encoding="utf-8")
    sys.stdout.reconfigure(encoding="utf-8", newline="\n")
    for line in sys.stdin:
        if not line.strip():
            continue
        request = json.loads(line)
        response = {
            "id": request["id"],
            "result": compute_project_diff(request["before"], request["after"]),
        }
        print(json.dumps(response, ensure_ascii=False, separators=(",", ":")), flush=True)


if __name__ == "__main__" and "--json-lines" in sys.argv:
    _json_lines()

File: src/v100/project-diff.mjs
/* Policy-driven semantic project diff and dependency closure. */
import DIFF_POLICY from "../../schemas/diff-policy.json" with { type: "json" };

import {
  digestCanonicalJson,
  evaluateProjectContract,
} from "./project-contract.mjs";

const clone = (value) => JSON.parse(digestCanonicalJson(value).canonical);
const pointerPart = (value) => String(value).replace(/~/g, "~0").replace(/\//g, "~1");
const compareText = (left, right) => left < right ? -1 : left > right ? 1 : 0;
const identityCollections = new Set(DIFF_POLICY.order.identity_keyed_collections);
const moveFields = new Set(DIFF_POLICY.category_rules.move_fields);

function validDocument(rawInput, label) {
  const evidence = evaluateProjectContract(rawInput);
  if (!evidence.valid) {
    const details = evidence.errors.map((error) => `${error.code}@${error.path}`).join(", ");
    throw new Error(`${label} project contract is invalid: ${details}`);
  }
  return { document: JSON.parse(evidence.canonical), evidence };
}

function valueHash(value, present) {
  return present ? digestCanonicalJson(value).digest : null;
}

function categoryFor(parts) {
  const last = parts.at(-1) || "";
  if (parts.some((part) => moveFields.has(part))) return "move";
  if (
    ["entity", "entity_id", "state_entity"].includes(last)
    || parts[0] === "bindings"
    || (parts[0] === "fields" && parts.length >= 3 && last === "entity")
    || (parts[0] === "slots" && parts.length >= 3 && last === "entity_id")
  ) return "binding";
  return DIFF_POLICY.category_rules.fallback_category;
}

function impactFor(category, path) {
  if (/^\/(security|permissions|plugins)(\/|$)/.test(path)) {
    return { severity: "critical", areas: ["security"] };
  }
  const areas = {
    add: ["none"],
    remove: ["operational", "referential"],
    move: ["visual"],
    binding: ["binding", "operational"],
    config: ["operational"],
  }[category];
  return {
    severity: DIFF_POLICY.impact.default_by_category[category],
    areas,
  };
}

function operation(category, path, collection, objectId, relativeParts, before, beforePresent, after, afterPresent) {
  return {
    id: `${category}:${path}`,
    category,
    path,
    collection,
    object_id: objectId,
    field: relativeParts.length ? `/${relativeParts.map(pointerPart).join("/")}` : "",
    before_hash: valueHash(before, beforePresent),
    after_hash: valueHash(after, afterPresent),
    before: beforePresent ? clone(before) : null,
    after: afterPresent ? clone(after) : null,
    impact: impactFor(category, path),
    requires: [],
  };
}

function compareValue(operations, before, beforePresent, after, afterPresent, context) {
  if (beforePresent && afterPresent && Object.is(before, after)) return;
  if (!beforePresent || !afterPresent) {
    const category = context.relativeParts.length === 0
      ? (beforePresent ? "remove" : "add")
      : categoryFor(context.relativeParts);
    operations.push(operation(
      category,
      context.path,
      context.collection,
      context.objectId,
      context.relativeParts,
      before,
      beforePresent,
      after,
      afterPresent,
    ));
    return;
  }
  const beforeArray = Array.isArray(before);
  const afterArray = Array.isArray(after);
  const beforeObject = before !== null && typeof before === "object";
  const afterObject = after !== null && typeof after === "object";
  if (beforeArray !== afterArray || beforeObject !== afterObject || !beforeObject || before === null || after === null) {
    operations.push(operation(
      categoryFor(context.relativeParts),
      context.path,
      context.collection,
      context.objectId,
      context.relativeParts,
      before,
      true,
      after,
      true,
    ));
    return;
  }
  if (beforeArray) {
    const length = Math.max(before.length, after.length);
    for (let index = 0; index < length; index += 1) {
      compareValue(operations, before[index], index < before.length, after[index], index < after.length, {
        ...context,
        path: `${context.path}/${index}`,
        relativeParts: [...context.relativeParts, String(index)],
      });
    }
    return;
  }
  const keys = [...new Set([...Object.keys(before), ...Object.keys(after)])].sort();
  for (const key of keys) {
    compareValue(
      operations,
      before[key],
      Object.hasOwn(before, key),
      after[key],
      Object.hasOwn(after, key),
      {
        ...context,
        path: `${context.path}/${pointerPart(key)}`,
        relativeParts: [...context.relativeParts, key],
      },
    );
  }
}

function identityMap(document, collection) {
  return new Map((document[collection] || []).map((entry) => [entry[DIFF_POLICY.identity_fields[collection]], entry]));
}

function addDependencies(operations, before, after) {
  const operationIds = new Set(operations.map(({ id }) => id));
  const beforeMaps = new Map();
  const afterMaps = new Map();
  for (const collection of Object.keys(DIFF_POLICY.identity_fields)) {
    beforeMaps.set(collection, identityMap(before, collection));
    afterMaps.set(collection, identityMap(after, collection));
  }
  for (const current of operations) {
    if (!current.collection || !current.object_id) continue;
    const references = DIFF_POLICY.dependencies.references.filter(({ from }) => from === current.collection);
    const sourceMap = current.category === "remove" ? beforeMaps : afterMaps;
    const source = sourceMap.get(current.collection)?.get(current.object_id);
    if (!source) continue;
    const requirements = new Map();
    for (const reference of references) {
      for (const field of reference.fields) {
        const targetId = source[field];
        if (typeof targetId !== "string") continue;
        const targetPath = `/${reference.to}/${pointerPart(targetId)}`;
        const targetOperation = current.category === "remove" ? `remove:${targetPath}` : `add:${targetPath}`;
        if (operationIds.has(targetOperation)) {
          requirements.set(targetOperation, {
            operation_id: targetOperation,
            reason: `reference:${reference.from}.${field}->${reference.to}`,
          });
        }
      }
    }
    current.requires = [...requirements.values()].sort((left, right) => compareText(left.operation_id, right.operation_id));
  }
}

export function computeProjectDiff(beforeInput, afterInput) {
  const beforeResult = validDocument(beforeInput, "source");
  const afterResult = validDocument(afterInput, "candidate");
  const before = beforeResult.document;
  const after = afterResult.document;
  const operations = [];
  const orderingNoise = [];
  const keys = [...new Set([...Object.keys(before), ...Object.keys(after)])].sort();
  for (const key of keys) {
    if (identityCollections.has(key) && Array.isArray(before[key]) && Array.isArray(after[key])) {
      const beforeMap = identityMap(before, key);
      const afterMap = identityMap(after, key);
      const beforeIds = [...beforeMap.keys()];
      const afterIds = [...afterMap.keys()];
      if (beforeIds.length === afterIds.length
        && beforeIds.every((id) => afterMap.has(id))
        && beforeIds.some((id, index) => id !== afterIds[index])) {
        orderingNoise.push(`/${pointerPart(key)}`);
      }
      const ids = [...new Set([...beforeIds, ...afterIds])].sort();
      for (const id of ids) {
        compareValue(
          operations,
          beforeMap.get(id),
          beforeMap.has(id),
          afterMap.get(id),
          afterMap.has(id),
          { path: `/${pointerPart(key)}/${pointerPart(id)}`, collection: key, objectId: id, relativeParts: [] },
        );
      }
      continue;
    }
    compareValue(
      operations,
      before[key],
      Object.hasOwn(before, key),
      after[key],
      Object.hasOwn(after, key),
      { path: `/${pointerPart(key)}`, collection: null, objectId: null, relativeParts: [key] },
    );
  }
  operations.sort((left, right) => compareText(left.id, right.id));
  addDependencies(operations, before, after);
  return {
    policy_version: DIFF_POLICY.policy_version,
    source_digest: beforeResult.evidence.digest,
    candidate_digest: afterResult.evidence.digest,
    operations,
    ordering_noise: orderingNoise.sort(),
  };
}

export function expandDiffSelection(diffResult, selectedOperationIds) {
  const operations = new Map((diffResult?.operations || []).map((operationValue) => [operationValue.id, operationValue]));
  const requested = [...new Set(selectedOperationIds || [])].sort();
  for (const operationId of requested) {
    if (!operations.has(operationId)) throw new Error(`unknown selected operation ${operationId}`);
  }
  const state = new Map();
  const included = new Set();
  const added = new Map();
  const visit = (operationId, requiredBy = null, reason = null, chain = []) => {
    if (!operations.has(operationId)) throw new Error(`missing dependency operation ${operationId}`);
    if (state.get(operationId) === "visiting") {
      throw new Error(`cyclic diff dependency: ${[...chain, operationId].join(" -> ")}`);
    }
    if (state.get(operationId) === "done") return;
    state.set(operationId, "visiting");
    if (requiredBy !== null && !requested.includes(operationId) && !added.has(operationId)) {
      added.set(operationId, { operation_id: operationId, required_by: requiredBy, reason });
    }
    const requirements = [...(operations.get(operationId).requires || [])]
      .sort((left, right) => compareText(left.operation_id, right.operation_id));
    for (const requirement of requirements) {
      visit(requirement.operation_id, operationId, requirement.reason, [...chain, operationId]);
    }
    state.set(operationId, "done");
    included.add(operationId);
  };
  for (const operationId of requested) visit(operationId);
  return {
    selected: [...included].sort(),
    requested,
    added: [...added.values()].sort((left, right) => compareText(left.operation_id, right.operation_id)),
  };
}
